receiver/awsxrayreceiver/internal/udppoller/poller_test.go

```diff
@@ -0,0 +1,403 @@
+			lastEntry.Context[0].Type == zapcore.ErrorType &&
+			errors.As(lastEntry.Context[0].Interface.(error), &errRecv) &&
+			errors.Unwrap(lastEntry.Context[0].Interface.(error)).Error() == randErrStr.String()
+	}, "poller should encounter net read error")
+}
+
+func TestSocketGenericReadError(t *testing.T) {
+	_, p, recordedLogs := createAndOptionallyStartPoller(t, false)
+	// close the actual socket because we are going to mock it out below
+	p.(*poller).udpSock.Close()
+
+	randErrStr, _ := uuid.NewRandom()
+	p.(*poller).udpSock = &mockSocketConn{
+		expectedOutput: []byte("dontCare"),
+		expectedError: &mockGenericErr{
+			mockErrStr: randErrStr.String(),
+		},
+	}
+
+	p.Start(context.Background())
+
+	testutil.WaitFor(t, func() bool {
+		logs := recordedLogs.All()
+		lastEntry := logs[len(logs)-1]
+		var errRecv *internalErr.ErrRecoverable
+		return strings.Contains(lastEntry.Message, "recoverable socket read error") &&
+			lastEntry.Context[0].Type == zapcore.ErrorType &&
+			errors.As(lastEntry.Context[0].Interface.(error), &errRecv) &&
+			errors.Unwrap(lastEntry.Context[0].Interface.(error)).Error() == randErrStr.String()
+	}, "poller should encounter generic socket read error")
+}
+
+type mockNetError struct {
+	mockErrStr string
+	temporary  bool
+}
+
+func (m *mockNetError) Error() string {
+	return m.mockErrStr
+}
+
+func (m *mockNetError) Timeout() bool {
+	return false
+}
+
+func (m *mockNetError) Temporary() bool {
+	return m.temporary
+}
+
+type mockGenericErr struct {
+	mockErrStr string
+}
+
+func (m *mockGenericErr) Error() string {
+	return m.mockErrStr
+}
+
+type mockSocketConn struct {
+	expectedOutput []byte
+	expectedError  error
+	readCount      int
+	mu             sync.Mutex
+}
+
+func (m *mockSocketConn) Read(b []byte) (int, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	copied := copy(b, m.expectedOutput)
+	if m.readCount > 0 {
+		// intentionally slow to prevent a busy loop during any unit tests
+		// that involve the poll() function
+		time.Sleep(5 * time.Second)
+	}
+	m.readCount++
+	return copied, m.expectedError
+}
+
+func (m *mockSocketConn) Close() error { return nil }
+
+func createAndOptionallyStartPoller(t *testing.T, start bool) (string, Poller, *observer.ObservedLogs) {
+	addr, err := findAvailableAddress()
+	assert.NoError(t, err, "there should be address available")
+
+	logger, recorded := logSetup()
+	poller, err := New(&Config{
+		Transport:          Transport,
+		Endpoint:           addr,
+		NumOfPollerToStart: 2,
+	}, logger)
+	assert.NoError(t, err, "receiver should be created")
+
+	if start {
+		poller.Start(context.Background())
+	}
+	return addr, poller, recorded
+}
+
+// findAvailableAddress finds an available local address+port and returns it.
+// There might be race condition on the address returned by this function if
+// there's some other code that grab the address before we can listen on it.
+func findAvailableAddress() (string, error) {
```

Comment:
There are already standard helpers to do this, could you please use them?

Reply:
Ha yea I did find https://pkg.go.dev/go.opentelemetry.io/collector@v0.5.1-0.20200722180048-c0b3cf61a63a/testutil?tab=doc#GetAvailableLocalAddress. Unfortunately, this receiver needs UDP not TCP.

Ideally, it would be better to contribute to the collector repo so that testutil supports udp as well, then uses it in this PR. However, our timeline is a bit tight so would it be ok if I leave this PR as it is so that following work on this receiver is unblocked while I can create a separate PR to update `testutil` with UDP functionalities? Basically do things in parallel to speed things up :)

Reply:
You just need the port and then build the address from that so GetAvailablePort should be all that you need. That said it is a nit.

Reply:
`GetAvailablePort` also calls `GetAvailableLocalAddress` internally so it's probably back to the same problem. Noted that this is a nit, I will still put a backlog item on my side to contribute modification to the `testutil` for testing UDP-based plugins in the future.

Thanks!

Reply:
The helper functions are basically just to get an available port the fact that they use "tcp" when finding an available one doesn't matter in this context. Anyway, it is not a blocker.